File: design_patterns/C++/Behavioral/iterator.cxx
#include <iostream>
#include <vector>
#include <cmath>

class RadioStation {
  protected:
	float frequency = 0.0;

  public:
	RadioStation(float frequency) {
		this->frequency = frequency;
	}

	float getFrequency() {
		return this->frequency;
	}
};

class StationList {
  protected:
	std::vector<RadioStation*> stations;
	int counter = 0;

  public:
	void addStation(RadioStation* station) {
		this->stations.push_back(station);
	}

	void remove(RadioStation* toRemove) {
		float toRemoveFrequency = toRemove->getFrequency();
		auto newEnd = std::remove_if(this->stations.begin(),
		this->stations.end(), [&toRemoveFrequency](RadioStation * station) {
			return fabs(station->getFrequency() - toRemoveFrequency)
			       <= std::numeric_limits<float>::epsilon();
		});
		this->stations.erase(newEnd, this->stations.end());
	}

	int count() {
		return this->stations.size();
	}

	RadioStation* current() {
		return this->stations[this->counter];
	}

	int key() {
		return this->counter;
	}

	void next() {
		++this->counter;
	}

	void rewind() {
		this->counter = 0;
	}

	bool valid() {
		return this->counter < this->count();
	}
};

int main() {
	auto stationList = new StationList();
	stationList->addStation(new RadioStation(89));
	stationList->addStation(new RadioStation(101));
	stationList->addStation(new RadioStation(102));
	stationList->addStation(new RadioStation(103.2));
	for (auto i = stationList->count(); i > 0; --i) {
		std::cout << stationList->current()->getFrequency() << std::endl;
		stationList->next();
	}
	// =====
	// 89
	// 101
	// 102
	// 103.2
	// =====
	stationList->remove(new RadioStation(89));
	stationList->rewind();
	for (auto i = stationList->count(); i > 0; --i) {
		std::cout << stationList->current()->getFrequency() << std::endl;
		stationList->next();
	}
	// =====
	// 101
	// 102
	// 103.2
	// =====
	return 0;
}

File: design_patterns/Python/Behavioral/iterator.py
class RadioStation:

    def __init__(self, frequency):
        self.frequency = frequency

    def get_frequency(self):
        return self.frequency


class StationList:

    def __init__(self):
        self.stations = []
        self.counter = 0

    def add_station(self, station):
        self.stations.append(station)

    def remove_station(self, to_remove):
        to_remove_frequency = to_remove.get_frequency()
        self.stations = [s for s in self.stations if
                         s.get_frequency() != to_remove_frequency]

    def count(self):
        return len(self.stations)

    def current(self):
        return self.stations[self.counter]

    def key(self):
        return self.counter

    def next(self):
        self.counter += 1

    def rewind(self):
        self.counter = 0


if __name__ == '__main__':

    station_list = StationList()
    station_list.add_station(RadioStation(89))
    station_list.add_station(RadioStation(101))
    station_list.add_station(RadioStation(102))
    station_list.add_station(RadioStation(103.2))

    for i in range(station_list.count(), 0, -1):
        print(station_list.current().get_frequency())
        station_list.next()

    # 89
    # 101
    # 102
    # 103.2

    station_list.remove_station(RadioStation(89))
    station_list.rewind()

    for i in range(station_list.count(), 0, -1):
        print(station_list.current().get_frequency())
        station_list.next()

    # 101
    # 102
    # 103.2
